larpandoracontent/LArReclustering/CheatedThreeDClusteringTool.cc

```diff
@@ -0,0 +1,109 @@
+/**
+ *  @file   larpandoracontent/LArReclustering/CheatedThreeDClusteringTool.cc
+ *
+ *  @brief  Implementation file for the reclustering algorithm that uses transverse calorimetric profiles.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+
+#include "larpandoracontent/LArReclustering/CheatedThreeDClusteringTool.h"
+
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+CheatedThreeDClusteringTool::CheatedThreeDClusteringTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool CheatedThreeDClusteringTool::Run(const Algorithm *const pAlgorithm, std::vector<pandora::CaloHitList*> &newCaloHitListsVector)
+{
+    if (newCaloHitListsVector.empty() || newCaloHitListsVector.size() != 1)
+        return false;
+
+    CaloHitList *initialCaloHitList = newCaloHitListsVector.at(0);
+    newCaloHitListsVector.clear();
+    for (const CaloHit *const pCaloHit : *initialCaloHitList)
+    {
+        const CaloHit *const pParentCaloHit = static_cast<const CaloHit *>(pCaloHit->GetParentAddress());
+
+        int mainMcParticleIndex = this->GetMainMcParticleIndex(pAlgorithm, pParentCaloHit);
+
+        // Flag to indicate if a matching list was found
+        bool found = false;
+
+        // Search for an existing list with matching MC particle index
+        for (CaloHitList *caloHitList : newCaloHitListsVector)
+        {
+            const CaloHit *const pListParentCaloHit = static_cast<const CaloHit *>(caloHitList->front()->GetParentAddress());
+            int listMainMcParticleIndex = this->GetMainMcParticleIndex(pAlgorithm, pListParentCaloHit);
+            if (listMainMcParticleIndex == mainMcParticleIndex)
+            {
+                found = true;
+                caloHitList->push_back(pCaloHit);
+                break;
+            }
+        }
+
+        // If a list with matching MC particle hits is not found, create one
+        if (!found)
+        {
+            CaloHitList* newList = new CaloHitList();
+            newList->push_back(pCaloHit);
+            newCaloHitListsVector.push_back(newList);
+        }
+    }
+
+    return true;
+}
+
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+int CheatedThreeDClusteringTool::GetMainMcParticleIndex(const Algorithm *const pAlgorithm, const pandora::CaloHit *const pCaloHit)
+{
+    if (!pCaloHit) {
+        std::cerr << "pCaloHit is null!" << std::endl;
+        return -1;
+    }
+    const MCParticleList *pMCParticleList(nullptr);
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*pAlgorithm, m_mcParticleListName, pMCParticleList));
+    if (!pMCParticleList) {
+        std::cerr << "MCParticleList is null!" << std::endl;
+        return -1;
+    }
```

Comment:
This is just a nitpick, since we arguably aren't very consistent with this anywhere.... Could be nice to have the algorithm name in the error message, to make it clear where it comes from.

Reply:
I've added this, thanks!